virtcontainers/qemu.go

```diff
@@ -1191,6 +1229,42 @@ func (q *qemu) disconnect() {
 	q.qmpShutdown()
 }
 
+func (q *qemu) resizeMemory(reqMemMB uint32, memoryBlockSizeMB uint32) (uint32, error) {
+	currentMemory := q.config.MemorySize + uint32(q.state.HotpluggedMemory)
+	err := q.qmpSetup()
+	if err != nil {
+		return 0, err
+	}
+	if err := q.updateMemoryBalloon(reqMemMB); err != nil {
```

Comment:
Should `updateMemoryBalloon` be called after memory is hotplugged? For instance if we have 1G memory in vm now, we balloon 2G memory which exceed the total memory of vm, before we hotplug another 1G to vm. Is this feasible?

Reply:
As I know ballooning is effective only when the scope of memory resizing do not exceed the `VM max memory`. But I am not sure whether ballooning is constrained to the `VM max memory` originates from the VM’s creation or the `VM max memory` when we hotplug more memory to VM. Please take a look at this.

Reply:
> Should updateMemoryBalloon be called after memory is hotplugged? For instance if we have 1G memory in vm now, we balloon 2G memory which exceed the total memory of vm, before we hotplug another 1G to vm. Is this feasible?

Yes feasible, it is feasible, you can increase the ballon with more memory than you actually have in the VM, in that case the memory VM will be the `min(ballon, guest physical memory)`.

> As I know ballooning is effective only when the scope of memory resizing do not exceed the VM max memory.But I am not sure whether ballooning is constrained to the VM max memory originates from the VM’s creation or the VM max memory when we hotplug more memory to VM. Please take a look at this.

From my testing, if the balloon is limited, the max memory is not. So we are safe to hotplug more memory and the balloon should not be affected. The slot enabling is independent  in the of the balloon mechanism. We can change the order if you want.

Reply:
> From my testing, if the balloon is limited, the max memory is not. So we are safe to hotplug more memory and the balloon should not be affected. The slot enabling is independent in the of the balloon mechanism. We can change the order if you want.

For instance, if we have a VM with 1G memory, then we want to hotplug another 1G memory, we will balloon 2G and then hotplug 1G memory with qemu cmd. It will success when we balloon 2G, but in fact it will be 1G according to `min(ballon, guest physical memory)`. Finally, we will get a VM with 1G balloon, 2G physical memory. So I think we should change the order. 

But I am afraid although we hotplug 1G memory then balloon 2G, the balloon will still be 1G, as I am not sure whether ballooning is constrained to the VM max memory originates from the VM’s creation or the VM max memory when we hotplug more memory to VM.